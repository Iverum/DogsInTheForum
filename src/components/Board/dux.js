// Actions
const ADD_THREAD = 'ditf/boards/ADD_THREAD'
const CLEAR_THREADS = 'ditf/boards/CLEAR_THREADS'

// Reducer
export default function reducer(state = [], action = {}) {
  switch (action.type) {
    case ADD_THREAD: {
      return [...state, action.data]
    }

    case CLEAR_THREADS: {
      return []
    }

    default: {
      return state
    }
  }
}

// Action creators
export function addThread(thread) {
  return {
    type: ADD_THREAD,
    data: thread
  }
}

export function clear() {
  return { type: CLEAR_THREADS }
}
