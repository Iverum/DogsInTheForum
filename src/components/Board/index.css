.pagination {
  text-align: center;
}

.pagination li {
  display: inline-block;
  margin-left: 0.75rem;
  margin-right: 0.75rem;
}

.pagination li a {
  text-decoration: none;
}

.pagination .disabled {
  display: none;
}

.pagination .active a {
  font-weight: bolder;
}
