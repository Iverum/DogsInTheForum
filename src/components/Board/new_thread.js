import React, { Component, PropTypes } from 'react'
import Modal from 'react-modal'

const customStyles = {
  content : {
    top                   : '50%',
    left                  : '50%',
    right                 : 'auto',
    bottom                : 'auto',
    marginRight           : '-50%',
    transform             : 'translate(-50%, -50%)'
  }
}

export default class NewThreadDialog extends Component {
  constructor(props) {
    super(props)
    this.createThread = this.createThread.bind(this)
    this.startCreatingThread = this.startCreatingThread.bind(this)
    this.closeModal = this.closeModal.bind(this)
    this.onThreadNameChange = this.onThreadNameChange.bind(this)
    this.state = {
      threadName: ''
    }
  }

  startCreatingThread() {
    this.setState({
      threadName: ''
    })
  }

  closeModal() {
    this.setState({
      threadName: ''
    })
    this.props.onRequestClose()
  }

  onThreadNameChange(e) {
    this.setState({ threadName: e.target.value })
  }

  createThread() {
    this.props.onSubmit(this.state.threadName)
    this.closeModal()
  }

  render() {
    return (
      <Modal
        contentLabel='New thread name?'
        isOpen={this.props.isVisible}
        onRequestClose={this.props.onRequestClose}
        style={customStyles}
      >
        <h1>Create New Thread</h1>
        <label htmlFor="threadName">Thread name</label>
        <input
          className="u-full-width"
          type="text"
          placeholder="My awesome thread"
          value={this.state.threadName}
          onChange={this.onThreadNameChange}
          id="threadName"
        />
        <input
          className="button-primary u-full-width"
          type="button"
          value="Create Thread"
          onClick={this.createThread}
        />
      </Modal>
    )
  }

}

NewThreadDialog.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onRequestClose: PropTypes.func.isRequired
}
