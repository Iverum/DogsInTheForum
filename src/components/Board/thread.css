.title {
  font-size: 3.2rem;
}

.post-count {
  color: grey;
  font-size: 3.2rem;
  font-weight: 100;
  text-align: end;
}
