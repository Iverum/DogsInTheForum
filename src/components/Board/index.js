import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import { ListView, ListRows, Pagination } from 'react-list-combo'
import firebase from 'firebase'
import moment from 'moment'
import uuid from 'uuid/v4'
import './index.css'
import Thread from './thread'
import NewThread from './new_thread'
import * as boardActions from './dux'
import * as userActions from '../User/dux'

export class Board extends Component {

  constructor(props) {
    super(props)
    this.updateThreadsFromDatabase = this.updateThreadsFromDatabase.bind(this)
    this.startCreatingThread = this.startCreatingThread.bind(this)
    this.createThread = this.createThread.bind(this)
    this.endCreatingThread = this.endCreatingThread.bind(this)
    this.state = { creatingThread: false }
  }

  componentWillMount() {
    this.boardRef = firebase.database().ref('threads')
    this.boardRef.off()
    this.boardRef.on('child_added', this.updateThreadsFromDatabase)
    this.boardRef.on('child_changed', this.updateThreadsFromDatabase)
  }

  componentWillUnmount() {
    this.boardRef.off()
    this.boardRef = null
    this.props.dispatch(boardActions.clear())
  }

  startCreatingThread() {
    this.setState({
      creatingThread: true
    })
  }

  createThread(name) {
    const now = moment().format()
    const newThread = {
      uuid: uuid(),
      author: this.props.user.id,
      name,
      postCount: 0,
      created: now,
      updated: now
    }
    firebase.database().ref(`threads/${newThread.uuid}`).set(newThread)
  }

  endCreatingThread() {
    this.setState({
      creatingThread: false
    })
  }

  updateThreadsFromDatabase(data) {
    const thread = data.val()
    firebase.database().ref(`users/${thread.author}`).once('value', data => {
      this.props.dispatch(userActions.addUser(data.val()))
    })
    this.props.dispatch(boardActions.addThread(thread))
  }

  render() {
    const threads = this.props.threads.map(thread => {
      return {
        ...thread,
        author: this.props.users[thread.author].name
      }
    })

    return (
      <div className='twelve columns'>
        <NewThread
          isVisible={this.state.creatingThread}
          onSubmit={this.createThread}
          onRequestClose={this.endCreatingThread}
        />
        <header className="row">
          <h1 className="nine columns">Board</h1>
          <input
            className="button-primary three columns"
            type="button"
            value="Create Thread"
            onClick={this.startCreatingThread}
          />
        </header>
        <ListView
          initData={threads}
          perPage={9}
        >
          <ListRows><Thread /></ListRows>
          <Pagination className='pagination' />
        </ListView>
      </div>
    )
  }

}

Board.defaultProps = {
  threads: []
}

Board.propTypes = {
  threads: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    postCount: PropTypes.number.isRequired
  }))
}

export default connect(state => ({
  threads: state.board,
  user: state.users.currentUser,
  users: state.users.users
}))(Board)
