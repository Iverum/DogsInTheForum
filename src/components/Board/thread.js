import React, { Component, PropTypes } from 'react'
import './thread.css'

export default class Thread extends Component {

  render() {
    return (
      <div className="row">
        <a href='#' className='ten columns'><h2 className='title'>{this.props.data.name}</h2></a>
        <span className='two columns post-count'>{this.props.data.postCount}</span>
      </div>
    )
  }

}

Thread.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    postCount: PropTypes.number.isRequired
  })
}
