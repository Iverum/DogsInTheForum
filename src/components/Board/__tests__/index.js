import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Board from '../index'

test('board is rendered correctly', () => {
  const component = renderer.create(
    <Board />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
