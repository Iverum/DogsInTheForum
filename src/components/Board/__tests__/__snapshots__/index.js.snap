exports[`test board is rendered correctly 1`] = `
<div
  className="twelve columns">
  <noscript />
  <header
    className="row">
    <h1
      className="nine columns">
      Board
    </h1>
    <input
      className="button-primary three columns"
      onClick={[Function]}
      type="button"
      value="Create Thread" />
  </header>
  <div>
    <div
      className={undefined} />
  </div>
</div>
`;
